Test mission protocol with dropped messages

julianoes opened the pull request:
This adds dropping of incoming and outgoing messages for mission upload/download. The result is that the mission test breaks.

julianoes committed:
DronecodeSDK: update to v0.14.1

julianoes committed:
cmake: build DronecodeSDK with MavlinkPassthrough

julianoes committed:
mission: add MavlinkPassthrough plugin

julianoes committed:
param: the include path was fixed

julianoes committed:
mission: refactoring to run with drops

julianoes committed:
cmake: switch to temporary branch

We need a fix, otherwise messages are intercepted forever and not reset
properly.

julianoes committed:
tests: added class to simulate lossy links

This adds a class which takes care of the stuff to randomly drop a
certain percentage of the messages.

julianoes committed:
cmake: pin DronecodeSDK to hash

This includes the fix for reseting the intercepts.

julianoes committed:
lossy_link: make thread-safe

julianoes committed:
cmake: allow to use local DronecodeSDK artefacts

For debugging purposes it's handy to build and install the Dronecode
SDK manually.

julianoes committed:
README: document how to use custom DronecodeSDK

julianoes committed:
mission: don't clear for now

We don't clear the mission because that's not implemented yet.

bkueng commented on src/tests/lossy_link.cpp:
```diff
@@ -0,0 +1,10 @@
+    std::lock_guard<std::mutex> lock(_mutex);
+
+    std::uniform_real_distribution<> distribution(0.0f, 1.0f);
+    const float random_value = distribution(_engine);
```
I thought about this some more and I think it's better to use a separate engine (or `LossyLink` instance) for RX and TX, and init with a different seed.
The problem that can happen now is: whether a message is dropped or not depends on ordering of incoming vs outgoing messages. For example a random incoming message (e.g. heartbeat) might arrive just before we send a message, or it arrives directly after.
So there is timing-dependent behavior. Separating RX from TX will help making tests more reproducible.

We can increase reproducibility even more if we apply `drop` to only a specific set of incoming mavlink messages (e.g. all mission messages), so that a test is for example independent from streamed attitude messages.

julianoes replied:
Right, good point. I'll change it.

julianoes replied:
Done in f127017.

julianoes committed:
mission: specify drop rate in config

This moves the config for how many of the messages to drop to the
config.

julianoes committed:
mission: fix drop logic

We are supposed to return true to keep a message, not true to drop it.

julianoes committed:
mission: use two LossyLink instances

This way we should have more reproducible tests because the timing
matters less between incoming and outgoing messages.

julianoes committed:
mission: fix style

julianoes committed:
main: added test summary output

bkueng commented on README.md:
```diff
@@ -25,6 +25,19 @@ Before committing, make sure to run the code formatting and tidying using clang.
+export DRONECODE_SDK_INSTALL_DIR=$(pwd)/install
+make ENABLE_MAVLINK_PASSTHROUGH=1 INSTALL_PREFIX=$DRONECODE_SDK_INSTALL_DIR default install
+cd <wherever>/mavlink-testing-suite
+(mkdir -p build && cd build && cmake -DDRONECODE_SDK_INSTALL_DIR=$DRONECODE_SDK_INSTALL_DIR .. && make)
```
Why the brackets?

julianoes replied:
Because I don't want to end up in `build`, I want to be able to grep and edit the code without having to use `../`.

bkueng replied:
Fair point.

bkueng commented on src/main.cpp:
```diff
@@ -88,21 +93,48 @@ int main(int argc, char** argv)
+                      << std::endl;
+        }
+
+        // TODO: this is ugly but otherwise clang-tidy complains.
```
What's the complaint and how would you like it?

julianoes replied:
I used the tertiary operator and I got this weird complaint:
```
error: do not implicitly decay an array into a pointer; consider using gsl::array_view or an explicit cast instead [cppcoreguidelines-pro-bounds-array-to-pointer-decay,-warnings-as-errors]
        std::cout << "  => " << (fails[i] ? "failed" : "passed") << std::endl;
```

bkueng replied:
You can disable that warning here: https://github.com/Auterion/mavlink-testing-suite/blob/master/.clang-tidy#L1

julianoes replied:
Worked around: 7a06ffe007ec8ab1a8c71c54823431505c167bc4.

bkueng commented on src/tests/lossy_link.h:
```diff
@@ -0,0 +1,24 @@
+private:
+    std::mutex _mutex{};
+    // We need predictable results, so we always start with the same seed.
+    std::seed_seq _seed;
```
This doesn't have to be stored.

julianoes replied:
True, thx.

julianoes replied:
I couldn't do it :confused: .

bkueng replied:
What does that mean?

julianoes replied:
You have to help me with C++ :confused:

bkueng replied:
Here you go: https://github.com/Auterion/mavlink-testing-suite/pull/10/commits/6b566b9169cfe5e355e8e6f0d9dea4cfcdec59f7

bkueng commented on src/tests/lossy_link.cpp:
```diff
@@ -0,0 +1,13 @@
+
+bool LossyLink::drop(float ratio)
+{
+    std::lock_guard<std::mutex> lock(_mutex);
```
You don't need the lock anymore.

julianoes replied:
Well, not for our current usage but potentially still.

bkueng replied:
I wouldn't add precautionary locking, you could do that almost everywhere. We should do it only where needed, as it adds unnecessary overhead, both in code and runtime.

julianoes replied:
Ooookay.

julianoes replied:
Done in b0ac231ccd0db3a8e7fadad4784c6a4f21c3ef1a.

bkueng reviewed (commented):
A couple of details, looks good otherwise.

julianoes committed:
main: workaround clang-tidy warning

julianoes committed:
lossy_link: not thread-safe anymore

bkueng committed:
LossyLink: don't store _seed

julianoes commented on src/tests/lossy_link.cpp:
```diff
@@ -1,6 +1,6 @@
 #include "lossy_link.h"
 
-LossyLink::LossyLink(int seed_num) : _seed({seed_num}) {}
+LossyLink::LossyLink(int seed_num) : _engine(seed_num) {}
```
Oh, it's that easy? :smile: thanks!
